designs/plugins-phase-2.md

```diff
@@ -0,0 +1,105 @@
+# Plugins Phase 2
+
+## Overview
+
+Phase 2 of the scaffold plugin system adds support for chaining plugins together. The idea for plugins phase 2 is to have support for a plugin architecture that enables plugins to be separate binaries discoverable by the Kubebuilder CLI binary via user specified plugin file paths and that they are language indepedent. Some of the phase 2 high-level requirements summarized from the feature request and discussions are: 
+
+* mdbook-like structure
+    * A Plugin receives a JSON blob plugin "universe" from upstream plugins containing file "nodes", each of which contain file metadata.
+    * The plugin can add or delete nodes (nodes are not mutable), then returns the modified universe so downstream plugins can do the same.
+* Pre-generate a project file for handling complex plugin use-cases.
+* Enhanced "layout" PROJECT file key.
+    * Either an ordered list of plugin {name}/{version}, or ordered map of {name}/{version} to plugin-specific config.
+    * Order matters for certain plugins, ex. a bazel plugin would need to run after all Go files are generated.
```

Comment:
The current project file supports a comma-separated lists of plugins in the `layout` field and offers a unstructured field were plugins can store their config. While I do like the suggested approach of having an ordered map with plugin-keys to configs, that is a breaking change and would require project version 4. Project version 3 (current) is still able to support our needs so this can be implemented without having to wait for the fourth project version.